assets/js/wcpay-checkout.js

```diff
@@ -129,12 +129,103 @@ jQuery( function( $ ) {
 	var paymentMethodGenerated;
 
 	/**
-	 * Generates a payment method, saves its ID in a hidden input, and re-submits the form.
+	 * Saves the payment method ID in a hidden input, and re-submits the form.
 	 *
-	 * @param {object} $form The jQuery object for the form.
+	 * @param {object} $form         The jQuery object for the form.
+	 * @param {object} paymentMethod Payment method object.
+	 */
+	var handleOrderPayment = function( $form, paymentMethod ) {
+		var id = paymentMethod.id;
+
+		// Flag that the payment method has been successfully generated so that we can allow the form
+		// submission next time.
+		paymentMethodGenerated = true;
+
+		// Populate form with the payment method.
+		var paymentMethodInput = document.getElementById( 'wcpay-payment-method' );
+		paymentMethodInput.value = id;
+
+		// Re-submit the form.
+		$form.removeClass( 'processing' ).submit();
+	};
+
+	/**
+	 * Saves the setup intent ID in a hidden input, and re-submits the form.
+	 *
+	 * @param {object} $form       The jQuery object for the form.
+	 * @param {object} setupIntent Setup intent object.
+	 */
+	var submitFormWithSetupIntent = function( $form, setupIntent ) {
+		// Populate form with the setup intent.
+		var setupIntentInput = $( '<input type="hidden" id="wcpay-setup-intent" name="wcpay-setup-intent" />' );
+		setupIntentInput.val( setupIntent.id );
+		$form.append( setupIntentInput );
+
+		// Re-submit the form.
+		$form.removeClass( 'processing' ).submit();
+	};
+
+	/**
+	 * Creates and authorizes a setup intent, saves its ID in a hidden input, and re-submits the form.
+	 *
+	 * @param {object} $form         The jQuery object for the form.
+	 * @param {object} paymentMethod Payment method object.
+	 */
+	var handleAddCard = function( $form, paymentMethod ) {
+		/* eslint-disable camelcase */
+		$.post( wcpay_config.ajaxUrl, {
+			action: 'create_setup_intent',
+			'wcpay-payment-method': paymentMethod.id,
+			_ajax_nonce: wcpay_config.createSetupIntentNonce,
+		} )
+		/* eslint-enable camelcase */
+			.then( function( response ) {
+				if ( ! response.success ) {
+					return $.Deferred().reject( response.data.error );
+				}
+
+				var setupIntent = response.data;
+
+				if ( 'succeeded' === setupIntent.status ) {
```

Comment:
Those statuses would be handled by the Stripe JS `confirmCardSetup`.

If you'd like, we would be able to leave all statuses to be handled by Stripe JS by removing the [`confirm`](https://github.com/Automattic/woocommerce-payments/pull/758/files/d81aa9cd5cd30e126f41d2def3706f5b57990e4e#diff-0ea0ad0906fad7a8771a51086d62a812R265) parameter and calling `confirmCardSetup` for every setup intent. It would reduce the code complexity here at the expense of a delay when adding basic cards. It would probably be more resilient and aligned with Stripe's expected flow. What do you think?

Reply:
Good catch! That bug seems to happen when the form is resubmitted once the setup intent is confirmed. It happens because the jQuery UI `block` function is [called again by core](https://github.com/woocommerce/woocommerce/blob/master/assets/js/frontend/add-payment-method.js#L30). Luckily, we can fix it by overriding the default `ignoreIfBlocked` setting before submitting the form and I've fixed it in https://github.com/Automattic/woocommerce-payments/pull/758/commits/da3360e9a0343647e9ba84adf4ac5181c4f299c5.

I don't think removing the confirmation on setup intent creation would complicate somewhere else, as that bug would need to be fixed anyway and it was due to the `blocked` class being applied twice. So, I've removed that special `succeeded` case and left all confirmation attempts to Stripe JS in https://github.com/Automattic/woocommerce-payments/pull/758/commits/81bea86ce242e51200e920c66ceaa11647065fea.